ledger-core/ledger/server/buildersvc/bundle/writer.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package bundle
+
+import (
+	"sync"
+
+	"github.com/insolar/assured-ledger/ledger-core/ledger"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/synckit"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/throw"
+)
+
+func NewWriter(snap SnapshotWriter) Writer {
+	if snap == nil {
+		panic(throw.IllegalValue())
+	}
+	return &bundleWriter{snap: snap}
+}
+
+var _ Writer = &bundleWriter{}
+
+// bundleWriter implements necessary synchronization guarantees for the snapshot writer.
+type bundleWriter struct {
+	mutex sync.Mutex
+	snap  SnapshotWriter
+
+	lastReady synckit.SignalChannel
+}
+
+func (p *bundleWriter) WaitWriteBundles(done synckit.SignalChannel, fn func(bool)) bool {
+	p.mutex.Lock()
+	prev, next := p._prepareWait(fn != nil)
+	p.mutex.Unlock()
+
+	if next == nil {
+		// there is nothing to wait: (1) no writers (2) all done (3) already rolled back
+		// and there is no fn to call
+		return true
+	}
+
+	ok := false
+	if prev != nil {
+		select {
+		case _, ok = <- prev:
+		case <- done:
+			// make sure that commit/rollback status will be propagated properly
+			go propagateReady(prev, next)
+			if fn != nil {
+				fn(false)
+			}
+			return false
+		}
+	}
+
+	switch {
+	case fn == nil:
+		if ok {
+			// propagate status properly as someone can be after this wait operation, hence can be affected by rollback
+			next <- struct{}{}
+		}
+		close(next)
+		return true
+
+	case ok:
+		defer func() {
+			next <- struct{}{}
+			close(next)
+		}()
+	default:
+		defer close(next)
+	}
+
+	fn(true)
```

Comment:
above block can be written differently:

```
defer close(next)
if ok {
    defer func() { next <- struct{}{} }()
}
if fn != nil {
    fn(true)
}
return true
```

or 

```
defer func() {
  if ok {
    next <- struct{}{}
  }
  close(next)
}()


if fn != nil {
    fn(true)
}
return true
```

Either I find more readable than above switch.

Reply:
fine by me :)

Reply:
but reworked it for clarity